consensus/src/consensus_observer/observer.rs

```diff
@@ -0,0 +1,379 @@
+    root: Arc<Mutex<LedgerInfoWithSignatures>>,
+    // pending execute/commit blocks, also buffers when in sync mode
+    pending_blocks: Arc<Mutex<BTreeMap<Round, (OrderedBlock, Option<CommitDecision>)>>>,
+    // execution client to buffer manager
+    execution_client: Arc<dyn TExecutionClient>,
+    // Indicate if it's in state sync mode, hold the task handle.
+    sync_handle: Option<DropGuard>,
+    // Sender to notify the observer state sync to `(Epoch, Round)` is done.
+    sync_notifier: tokio::sync::mpsc::UnboundedSender<(u64, Round)>,
+    // Reconfig event listener to reload on-chain configs.
+    reconfig_events: ReconfigNotificationListener<DbBackedOnChainConfig>,
+    // Maps between block id and its payload, the same as payload manager returns.
+    payload_store: Arc<Mutex<HashMap<HashValue, ObserverDataStatus>>>,
+    // Publisher to forward payload message.
+    publisher: Option<Publisher>,
+}
+
+impl Observer {
+    pub fn new(
+        root: LedgerInfoWithSignatures,
+        execution_client: Arc<dyn TExecutionClient>,
+        sync_notifier: tokio::sync::mpsc::UnboundedSender<(u64, Round)>,
+        reconfig_events: ReconfigNotificationListener<DbBackedOnChainConfig>,
+        publisher: Option<Publisher>,
+    ) -> Self {
+        Self {
+            epoch: root.commit_info().epoch(),
+            root: Arc::new(Mutex::new(root)),
+            pending_blocks: Arc::new(Mutex::new(BTreeMap::new())),
+            execution_client,
+            sync_handle: None,
+            sync_notifier,
+            reconfig_events,
+            payload_store: Arc::new(Mutex::new(HashMap::new())),
+            publisher,
+        }
+    }
+
+    fn last_block(&self) -> BlockInfo {
+        self.pending_blocks
+            .lock()
+            .last_key_value()
+            .as_ref()
+            .map_or_else(
+                || self.root.lock().commit_info().clone(),
+                |(_, (last_blocks, _))| last_blocks.blocks.last().unwrap().block_info(),
+            )
+    }
+
+    fn commit_callback(&self) -> StateComputerCommitCallBackType {
+        let root = self.root.clone();
+        let pending_blocks = self.pending_blocks.clone();
+        let payload_store = self.payload_store.clone();
+        Box::new(move |blocks, ledger_info: LedgerInfoWithSignatures| {
+            {
+                let mut payload_store = payload_store.lock();
+                for block in blocks.iter() {
+                    payload_store.remove(&block.id());
+                }
+            }
+            {
+                let mut pending_blocks = pending_blocks.lock();
+                *pending_blocks =
+                    pending_blocks.split_off(&(ledger_info.commit_info().round() + 1));
+            }
+            *root.lock() = ledger_info;
+        })
+    }
+
+    fn forward_decision(&self, decision: CommitDecision) {
+        let (tx, _rx) = oneshot::channel();
+        self.execution_client
+            // it's just a dummy rpc message
```

Comment:
Is this unnecessary?

Reply:
what do you mean? the rpc part? I thought it only accepts rpc but if direct send works we can use it

Reply:
nvm